File: backend/bedrock-question-service/evals/observe_html_artifact.py
"""Bounded parent transport for the app-owned HTML observer; never model JS.

The caller must durably record the prepared job before calling. Local process
termination cannot prove browser cleanup; any unconfirmed cleanup blocks use.
"""

import base64
import copy
import json
import os
from pathlib import Path
import selectors
import signal
import subprocess
import tempfile
import time

from evals.html_artifact_question import (
    OBSERVATION_PROTOCOL,
    OBSERVER,
    SETTINGS,
    canonical,
    digest,
)

PROCESS_TIMEOUT_SECONDS = 45
CAPTURE_BYTES = 32768
INPUT_BYTES = 16384
REAP_GRACE_SECONDS = 0.2


def _unique(pairs):
    result = {}
    for key, value in pairs:
        if key in result:
            raise ValueError("duplicate_json_key")
        result[key] = value
    return result


def _terminate(process):
    for sig in (signal.SIGTERM, signal.SIGKILL):
        try:
            os.killpg(process.pid, sig)
        except ProcessLookupError:
            pass
        try:
            process.wait(timeout=REAP_GRACE_SECONDS)
        except subprocess.TimeoutExpired:
            pass
    return process.returncode is not None


def _valid_envelope(envelope, job):
    """Validate transport evidence without requiring an offered answer match."""
    keys = {
        "protocol",
        "job_id",
        "artifact_sha256",
        "probe_sha256",
        "observer_source_sha256",
        "status",
        "reason",
        "browser",
        "settings",
        "values",
        "serialized_dom",
        "serialized_dom_sha256",
        "target_outer_html",
        "cleanup",
    }
    if (
        type(envelope) is not dict
        or set(envelope) != keys
        or envelope["protocol"] != OBSERVATION_PROTOCOL
        or any(
            envelope[k] != job[k]
            for k in (
                "job_id",
                "artifact_sha256",
                "probe_sha256",
                "observer_source_sha256",
            )
        )
        or canonical(envelope["settings"]) != canonical(SETTINGS)
        or type(envelope["reason"]) is not str
        or len(envelope["reason"]) > 160
        or type(envelope["browser"]) is not dict
        or set(envelope["browser"]) != {"name", "version"}
        or envelope["browser"]["name"] != "chromium"
        or type(envelope["browser"]["version"]) is not str
        or not 1 <= len(envelope["browser"]["version"].strip()) <= 80
        or len(envelope["browser"]["version"]) > 80
    ):
        return False
    if envelope["status"] == "unsupported":
        return (
            bool(envelope["reason"].strip())
            and all(
                envelope[k] == ""
                for k in (
                    "serialized_dom",
                    "serialized_dom_sha256",
                    "target_outer_html",
                )
            )
            and envelope["values"] is None
        )
    if envelope["status"] != "observed" or envelope["reason"] != "":
        return False
    return (
        type(envelope["values"]) is list
        and len(envelope["values"]) == len(job["spec"]["properties"])
        and all(type(v) is bool for v in envelope["values"])
        and all(
            type(envelope[k]) is str and 1 <= len(envelope[k]) <= 4096
            for k in ("serialized_dom", "target_outer_html")
        )
        and digest(envelope["serialized_dom"]) == envelope["serialized_dom_sha256"]
    )


def observe_html_job(job, *, node, playwright_module, browsers_path):
    """Return bounded raw transport evidence; caller separately prepares/binds."""
    started = time.monotonic()
    stdin_text = canonical(job) + "\n"
    data = stdin_text.encode("utf-8")
    record = {
        "request": copy.deepcopy(job),
        "stdin_text": stdin_text,
        "stdin_sha256": digest(stdin_text),
        "stdin_bytes": len(data),
        "observer_path": str(OBSERVER),
        "node": str(Path(node).resolve()),
        "status": "operational_failure",
        "reason": "",
        "returncode": None,
        "process_deadline_seconds": PROCESS_TIMEOUT_SECONDS,
        "capture_limit_bytes_per_stream": CAPTURE_BYTES,
        "cleanup_grace_seconds": REAP_GRACE_SECONDS * 3,
        "envelope": None,
        "child_started": False,
    }
    buffers = {name: bytearray() for name in ("stdout", "stderr")}
    seen = {name: 0 for name in buffers}
    process, failure, sent, terminated = None, "", 0, False
    selector = selectors.DefaultSelector()
    try:
        if len(data) > INPUT_BYTES:
            raise ValueError("input_byte_limit_exceeded")
        with tempfile.TemporaryDirectory(prefix="checkpoint-html-transport-") as temp:
            env = {
                "PATH": os.defpath,
                "TMPDIR": temp,
                "LANG": "C",
                "CHECKPOINT_PLAYWRIGHT_MODULE": str(Path(playwright_module).resolve()),
                "PLAYWRIGHT_BROWSERS_PATH": str(Path(browsers_path).resolve()),
            }
            if "HOME" in os.environ:
                env["HOME"] = os.environ["HOME"]
            record["environment"] = env
            process = subprocess.Popen(
                [record["node"], str(OBSERVER)],
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                env=env,
                cwd=temp,
                start_new_session=True,
            )
            record["child_started"] = True
            record["pid"] = process.pid
            for name in ("stdin", "stdout", "stderr"):
                stream = getattr(process, name)
                os.set_blocking(stream.fileno(), False)
                selector.register(
                    stream,
                    selectors.EVENT_WRITE if name == "stdin" else selectors.EVENT_READ,
                    name,
                )
            deadline = started + PROCESS_TIMEOUT_SECONDS
            while selector.get_map():
                remaining = deadline - time.monotonic()
                if remaining <= 0:
                    failure = "process_deadline"
                    break
                for key, _ in selector.select(min(remaining, 0.05)):
                    if key.data == "stdin":
                        try:
                            sent += os.write(key.fd, data[sent : sent + 4096])
                        except BrokenPipeError:
                            selector.unregister(key.fileobj)
                            key.fileobj.close()
                            continue
                        if sent == len(data):
                            selector.unregister(key.fileobj)
                            key.fileobj.close()
                        continue
                    chunk = os.read(key.fd, 8192)
                    if not chunk:
                        selector.unregister(key.fileobj)
                        key.fileobj.close()
                        continue
                    name = key.data
                    seen[name] += len(chunk)
                    buffers[name].extend(
                        chunk[: max(0, CAPTURE_BYTES - len(buffers[name]))]
                    )
                    if seen[name] > CAPTURE_BYTES:
                        failure = name + "_capture_limit"
                        break
                if failure:
                    break
            if not failure:
                try:
                    process.wait(timeout=max(0, deadline - time.monotonic()))
                except subprocess.TimeoutExpired:
                    failure = "process_deadline"
            if failure:
                terminated = True
                _terminate(process)
            # Drain remaining buffered output without retaining excess bytes or
            # waiting indefinitely on a descendant that inherited a pipe.
            drain_deadline = time.monotonic() + REAP_GRACE_SECONDS
            while selector.get_map() and time.monotonic() < drain_deadline:
                for key, _ in selector.select(0.01):
                    if key.data == "stdin":
                        selector.unregister(key.fileobj)
                        key.fileobj.close()
                        continue
                    chunk = os.read(key.fd, 8192)
                    if not chunk:
                        selector.unregister(key.fileobj)
                        key.fileobj.close()
                        continue
                    seen[key.data] += len(chunk)
                    buffers[key.data].extend(
                        chunk[: max(0, CAPTURE_BYTES - len(buffers[key.data]))]
                    )
    except (OSError, ValueError) as error:
        failure = (
            str(error)
            if str(error) == "input_byte_limit_exceeded"
            else type(error).__name__
        )
        if process is not None:
            terminated = True
            _terminate(process)
    finally:
        for key in list(selector.get_map().values()):
            key.fileobj.close()
        selector.close()
        if process is not None:
            record["returncode"] = process.returncode
        record["stdin_bytes_sent"] = sent
        record["cleanup"] = {
            "child_reaped": process is not None and process.returncode is not None,
            "termination_attempted": terminated,
            "browser_cleanup_confirmed": False,
        }
        record["elapsed_seconds"] = round(time.monotonic() - started, 6)
    for name, buffer in buffers.items():
        raw = bytes(buffer)
        record[name + "_base64"] = base64.b64encode(raw).decode("ascii")
        record[name + "_bytes_seen"] = seen[name]
        record[name + "_retained_bytes"] = len(raw)
        try:
            record[name] = raw.decode("utf-8")
        except UnicodeDecodeError:
            record[name] = None
            failure = failure or "invalid_output_utf8"
    try:
        envelope = json.loads(record["stdout"] or "", object_pairs_hook=_unique)
        record["envelope"] = envelope
        valid = _valid_envelope(envelope, job)
        closed = valid and envelope.get("cleanup") == {
            "context": "closed",
            "browser": "closed",
        }
        record["cleanup"]["browser_cleanup_confirmed"] = bool(closed and not terminated)
        if not failure and sent == len(data) and closed:
            if (
                record["returncode"] == 0
                and envelope.get("status") == "observed"
                and envelope.get("reason") == ""
            ):
                record["status"] = "completed"
            elif record["returncode"] == 1 and envelope.get("status") == "unsupported":
                record["status"] = "unsupported"
            else:
                failure = "observer_failed"
        else:
            failure = failure or "incomplete_or_mismatched_observation"
    except (ValueError, TypeError, KeyError, RecursionError):
        failure = failure or "malformed_observer_output"
    record["reason"] = failure
    return record

File: backend/bedrock-question-service/evals/observe_html_artifact.js
#!/usr/bin/env node
'use strict';
// One isolated native observation. All executed JavaScript below is app-owned.
// Input HTML/properties are data; there is no eval or model-supplied script API.
const fs = require('fs');
const crypto = require('crypto');
const PROTOCOL = 'checkpoint.html_artifact_job.v1';
const OBSERVATION_PROTOCOL = 'checkpoint.html_artifact_observation.v1';
const PROPERTIES = ['disabled', "matches(':disabled')", 'required', 'readOnly', 'willValidate', 'validity.valueMissing', 'validity.valid', 'checkValidity()'];
const SETTINGS = {offline: true, javaScriptEnabled: false, acceptDownloads: false, serviceWorkers: 'block', allRoutesAborted: true, chromiumSandbox: true};
const TAGS = ['form', 'fieldset', 'legend', 'label', 'input', 'button', 'textarea', 'select', 'option', 'optgroup', 'div', 'span', 'p', 'br'];
const ATTRIBUTES = ['id', 'type', 'value', 'name', 'disabled', 'required', 'readonly', 'checked', 'multiple', 'selected', 'min', 'max', 'step', 'minlength', 'maxlength', 'pattern', 'for', 'form', 'placeholder', 'label', 'size'];
const hash = value => crypto.createHash('sha256').update(value).digest('hex');
const length = value => [...value].length;
const canonical = value => Array.isArray(value) ? '[' + value.map(canonical).join(',') + ']' : value !== null && typeof value === 'object' ? '{' + Object.keys(value).sort().map(k => JSON.stringify(k) + ':' + canonical(value[k])).join(',') + '}' : JSON.stringify(value);
const fail = reason => { throw new Error(reason); };
const object = (value, keys) => { if (value === null || Array.isArray(value) || typeof value !== 'object' || canonical(Object.keys(value).sort()) !== canonical([...keys].sort())) fail('invalid_object_fields'); };
const text = (value, maximum, minimum = 1) => {
  if (typeof value !== 'string' || length(value.trim()) < minimum || length(value) > maximum || [...value].some(c => /[\p{Cs}\p{Cf}]/u.test(c) || (/\p{Cc}/u.test(c) && !'\n\r\t'.includes(c)))) fail('invalid_or_oversized_text');
};
function readInput() {
  const chunks = []; let bytes = 0;
  while (true) {
    const buffer = Buffer.alloc(4096), count = fs.readSync(0, buffer, 0, buffer.length, null);
    if (!count) break;
    bytes += count;
    if (bytes > 16384) fail('input_byte_limit_exceeded');
    chunks.push(buffer.subarray(0, count));
  }
  return JSON.parse(Buffer.concat(chunks).toString('utf8'));
}
function validateJob(job) {
  object(job, ['protocol', 'spec', 'prompt', 'choices', 'artifact_sha256', 'probe_sha256', 'observer_source_sha256', 'job_id']);
  const s = job.spec;
  object(s, ['html', 'targetID', 'properties', 'alternatives', 'explanation', 'topic', 'difficulty']);
  text(s.html, 320);
  if (typeof s.targetID !== 'string' || !/^[A-Za-z][A-Za-z0-9_-]{0,23}$/.test(s.targetID)) fail('invalid_target_id');
  if (!Array.isArray(s.properties) || s.properties.length < 2 || s.properties.length > 4 || s.properties.some(p => !PROPERTIES.includes(p)) || new Set(s.properties).size !== s.properties.length) fail('invalid_properties');
  if (!Array.isArray(s.alternatives) || s.alternatives.length !== 4) fail('four_alternatives_required');
  for (const alternative of s.alternatives) {
    object(alternative, ['values', 'reason']);
    if (!Array.isArray(alternative.values) || alternative.values.length !== s.properties.length || alternative.values.some(v => typeof v !== 'boolean')) fail('invalid_boolean_tuple');
    text(alternative.reason, 280, 12);
  }
  text(s.explanation, 420, 12); text(s.topic, 80);
  if (!Number.isInteger(s.difficulty) || s.difficulty < 1 || s.difficulty > 5) fail('invalid_proposed_difficulty');
  const prompt = `In a fresh Chromium document, for #${s.targetID}, what are ${s.properties.join(', ')} (in order)?\n${s.html}`;
  const choices = s.alternatives.map(a => '[' + a.values.join(', ') + ']');
  if (length(prompt) > 320 || choices.some(c => length(c) > 140) || new Set(choices).size !== 4) fail('display_limit_or_duplicate_choice');
  if (job.protocol !== PROTOCOL || job.prompt !== prompt || canonical(job.choices) !== canonical(choices) || job.artifact_sha256 !== hash(s.html) || job.probe_sha256 !== hash(canonical({targetID: s.targetID, properties: s.properties})) || job.observer_source_sha256 !== hash(fs.readFileSync(__filename))) fail('job_binding_mismatch');
  const unsigned = {...job}; delete unsigned.job_id;
  if (job.job_id !== hash(canonical(unsigned))) fail('job_id_mismatch');
}
function bounded(promise, milliseconds) {
  let timer;
  return Promise.race([promise, new Promise((_, reject) => { timer = setTimeout(() => reject(new Error('local_operation_deadline')), milliseconds); })]).finally(() => clearTimeout(timer));
}
async function main() {
  let job, browser, context;
  const result = {
    protocol: OBSERVATION_PROTOCOL, job_id: null, artifact_sha256: null, probe_sha256: null,
    observer_source_sha256: hash(fs.readFileSync(__filename)), status: 'operational_failure', reason: '',
    browser: {name: 'chromium', version: ''}, settings: SETTINGS, values: null,
    serialized_dom: '', serialized_dom_sha256: '', target_outer_html: '',
    cleanup: {context: 'not_created', browser: 'not_created'},
  };
  try {
    job = readInput(); validateJob(job);
    for (const key of ['job_id', 'artifact_sha256', 'probe_sha256']) result[key] = job[key];
    const {chromium} = require(process.env.CHECKPOINT_PLAYWRIGHT_MODULE || 'playwright');
    // A fresh managed browser, never a user's browser/profile. Do not pass AWS or
    // other application credentials into the browser process environment.
    const browserEnv = Object.fromEntries(['PATH', 'HOME', 'TMPDIR', 'LANG'].filter(k => process.env[k] !== undefined).map(k => [k, process.env[k]]));
    result.cleanup.browser = 'launch_unconfirmed';
    browser = await chromium.launch({headless: true, chromiumSandbox: true, timeout: 10000, env: browserEnv});
    result.cleanup.browser = 'open';
    const browserVersion = browser.version();
    text(browserVersion, 80);
    result.browser.version = browserVersion;
    result.cleanup.context = 'creation_unconfirmed';
    context = await bounded(browser.newContext({offline: true, javaScriptEnabled: false, acceptDownloads: false, serviceWorkers: 'block'}), 3000);
    result.cleanup.context = 'open';
    await bounded(context.route('**/*', route => route.abort()), 3000);
    const page = await bounded(context.newPage(), 3000);
    // Parse in an inert template first: unsupported elements/attributes never
    // enter the active document. The model cannot supply observer expressions.
    const passive = await bounded(page.evaluate(({html, tags, attributes}) => {
      const template = document.createElement('template');
      template.innerHTML = html;
      return [...template.content.querySelectorAll('*')].every(el => el.namespaceURI === 'http://www.w3.org/1999/xhtml' && tags.includes(el.localName) && [...el.attributes].every(a => attributes.includes(a.name)));
    }, {html: job.spec.html, tags: TAGS, attributes: ATTRIBUTES}), 3000);
    if (!passive) { result.status = 'unsupported'; fail('unsupported_tag_or_attribute'); }
    await page.setContent(job.spec.html, {timeout: 3000, waitUntil: 'domcontentloaded'});
    const observed = await bounded(page.evaluate(({targetID, properties, tags, attributes}) => {
      const allowed = [...document.querySelectorAll('*')].every(el => el.namespaceURI === 'http://www.w3.org/1999/xhtml' && (['html', 'head', 'body'].includes(el.localName) || tags.includes(el.localName)) && [...el.attributes].every(a => attributes.includes(a.name)));
      const ids = [...document.querySelectorAll('[id]')].map(el => el.id);
      const targets = [...document.querySelectorAll('[id]')].filter(el => el.id === targetID);
      if (!allowed || new Set(ids).size !== ids.length || targets.length !== 1) return {reason: 'unsupported_dom_or_target'};
      const target = targets[0];
      if (!['input', 'button', 'textarea', 'select'].includes(target.localName)) return {reason: 'unsupported_target_element'};
      const values = properties.map(property => {
        switch (property) {
          case 'disabled': return target.disabled;
          case "matches(':disabled')": return target.matches(':disabled');
          case 'required': return target.required;
          case 'readOnly': return target.readOnly;
          case 'willValidate': return target.willValidate;
          case 'validity.valueMissing': return target.validity.valueMissing;
          case 'validity.valid': return target.validity.valid;
          case 'checkValidity()': return target.checkValidity();
          default: return undefined;
        }
      });
      if (values.some(v => typeof v !== 'boolean')) return {reason: 'property_not_boolean_on_target'};
      return {values, targetOuterHTML: target.outerHTML, dom: document.documentElement.outerHTML};
    }, {targetID: job.spec.targetID, properties: job.spec.properties, tags: TAGS, attributes: ATTRIBUTES}), 3000);
    if (observed.reason) { result.status = 'unsupported'; fail(observed.reason); }
    if (length(observed.dom) > 4096 || length(observed.targetOuterHTML) > 4096) fail('native_capture_limit_exceeded');
    result.values = observed.values;
    result.serialized_dom = observed.dom;
    result.serialized_dom_sha256 = hash(observed.dom);
    result.target_outer_html = observed.targetOuterHTML;
    result.status = 'observed';
  } catch (error) {
    // Preserve bounded operational diagnostics, never environment or stack text.
    const known = /^(invalid_|unsupported_|missing_|duplicate_|unbalanced_|four_|display_|job_|input_|local_|native_|property_)/.test(error.message || '');
    result.reason = known ? String(error.message).slice(0, 160) : String(error.name || 'Error').slice(0, 80);
  } finally {
    for (const [name, resource] of [['context', context], ['browser', browser]]) {
      if (!resource) continue;
      try { await bounded(resource.close(), 3000); result.cleanup[name] = 'closed'; }
      catch (_) { result.cleanup[name] = 'failed'; result.status = 'operational_failure'; result.reason = 'cleanup_failed'; }
    }
  }
  let encoded = JSON.stringify(result) + '\n';
  if (Buffer.byteLength(encoded, 'utf8') > 32768) {
    result.status = 'operational_failure'; result.reason = 'output_byte_limit_exceeded';
    result.serialized_dom = ''; result.target_outer_html = ''; result.values = null;
    encoded = JSON.stringify(result) + '\n';
  }
  // An isolated single-job CLI. A parent should also enforce a process deadline;
  // a failed cleanup is explicit and must never produce an admitted question.
  process.stdout.write(encoded, () => process.exit(result.status === 'observed' ? 0 : 1));
}
main().catch(() => { process.exitCode = 1; });
